tests/unit/steps/test_hypervisor.py

```diff
@@ -14,56 +14,156 @@
+
+
+def test_generate_upgrade_steps():
+    """Test generating of upgrade steps."""
+    target = OpenStackRelease("victoria")
+    machines = [Machine(f"{i}", (), f"az{i}") for i in range(3)]
+    app_units = {
+        "app1": [Unit(f"app1/{i}", machines[i], "1") for i in range(3)],
+        "app2": [Unit(f"app2/{i}", machines[i], "1") for i in range(3)],
+    }
+    apps = [_generate_app("app1"), _generate_app("app2"), _generate_app("app3")]
+    exp_steps = apps[0].upgrade_steps.return_value + apps[1].upgrade_steps.return_value
+    # Note(rgildein): it contains only two apps, so app3 should be skipped
+    group = HypervisorGroup("test", app_units)
+    planner = HypervisorUpgradePlanner(apps, machines)
+
+    steps = planner._generate_upgrade_steps(target, False, group)
+
+    apps[0].upgrade_steps.assert_called_once_with(target, app_units["app1"], False)
+    apps[1].upgrade_steps.assert_called_once_with(target, app_units["app2"], False)
+    apps[2].upgrade_steps.assert_not_called()
+
+    assert steps == exp_steps
 
 
 def test_generate_post_upgrade_steps():
     """Test generating of post-upgrade steps."""
     target = OpenStackRelease("victoria")
-    units = ["1", "2", "3"]
     machines = [Machine(f"{i}", (), f"az{i}") for i in range(3)]
-    apps = [_generate_app() for _ in range(3)]
-    group = HypervisorGroup("test", {app.name.return_value: units for app in apps})
-
+    app_units = {
+        "app1": [Unit(f"app1/{i}", machines[i], "1") for i in range(3)],
+        "app2": [Unit(f"app2/{i}", machines[i], "1") for i in range(3)],
+    }
+    apps = [_generate_app("app1"), _generate_app("app2"), _generate_app("app3")]
+    exp_steps = apps[0].post_upgrade_steps.return_value + apps[1].post_upgrade_steps.return_value
+    # Note(rgildein): it contains only two apps, so app3 should be skipped
+    group = HypervisorGroup("test", app_units)
     planner = HypervisorUpgradePlanner(apps, machines)
+
     steps = planner._generate_post_upgrade_steps(target, group)
 
-    for step, app in zip(steps, apps[::-1]):  # using reversed order of applications
-        app.post_upgrade_steps.assert_called_once_with(target, units=units)
-        assert step == app.post_upgrade_steps.return_value[0]  # mocked app contain single step
+    apps[0].post_upgrade_steps.assert_called_once_with(target, units=app_units["app1"])
+    apps[1].post_upgrade_steps.assert_called_once_with(target, units=app_units["app2"])
+    apps[2].post_upgrade_steps.assert_not_called()
+
+    assert steps == exp_steps
+
+
+@patch("cou.steps.hypervisor.HypervisorUpgradePlanner.get_azs")
+@patch("cou.steps.hypervisor.HypervisorUpgradePlanner._upgrade_plan_sanity_checks")
+@patch("cou.steps.hypervisor.HypervisorUpgradePlanner._generate_pre_upgrade_steps")
+@patch("cou.steps.hypervisor.HypervisorUpgradePlanner._generate_upgrade_steps")
+@patch("cou.steps.hypervisor.HypervisorUpgradePlanner._generate_post_upgrade_steps")
+def test_generate_upgrade_plan(
```

Comment:
I'm missing an unit tests that shows the expect behavior when an `_upgrade_plan_sanity_checks` fails

Reply:
What is that? If `_upgrade_plan_sanity_checks` raise an exception, this function will fail an exception continues until we catch it. What do you want to test here?

Reply:
I do not see the point of that, since we are not catching any errors here. If any steps, e.g. _generate_pre_upgrade_steps failed, the whole function will fail.